front/pages/[user]/a/[sId]/execute/index.jsx

```diff
@@ -0,0 +1,403 @@
+            <ObjectViewer value={preprocessedOutput} />
+          ) : null}
+        </div>
+      ) : null}
+    </>
+  );
+}
+
+function ExecuteOutput({ executionLogs, expandedByBlockName, onToggleExpand }) {
+  return executionLogs.blockOrder.length ? (
+    <>
+      {executionLogs.blockOrder.map((blockName, index) => {
+        const lastEventForBlock = executionLogs.lastEventByBlockName[blockName];
+        const outputForBlock = executionLogs.outputByBlockName[blockName];
+        return (
+          <ExecuteOutputLine
+            key={blockName}
+            blockName={blockName}
+            outputForBlock={outputForBlock}
+            lastEventForBlock={lastEventForBlock}
+            expanded={expandedByBlockName[blockName]}
+            onToggleExpand={() => onToggleExpand(blockName)}
+          />
+        );
+      })}
+      <br /> {/*TODO: no br*/}
+    </>
+  ) : null;
+}
+
+function ExecuteInput({ inputName, inputValue, onChange }) {
+  return (
+    <div key={inputName} className="grid grid-cols-10">
+      <div className="flex group items-center bg-slate-300">
+        <input
+          className={classNames(
+            "px-1 py-1 font-normal text-[13px] font-mono bg-slate-300 border-0 outline-none focus:outline-none w-full",
+            "border-white ring-0 focus:ring-0 focus:border-white"
+          )}
+          readOnly={true}
+          value={inputName}
+        />
+      </div>
+      <div
+        className={classNames(
+          "col-span-7 inline-grid space-y-0 resize-none text-[13px] font-mono px-0 py-0 border bg-slate-100",
+          "border-slate-100"
+        )}
+      >
+        <TextareaAutosize
+          minRows={1}
+          className={classNames(
+            "w-full resize-none font-normal text-[13px] font-mono px-1 py-0 bg-transparent border-0 ring-0 focus:ring-0",
+            "text-gray-700"
+          )}
+          value={inputValue || ""}
+          onChange={(e) => onChange(e.target.value)}
+        />
+      </div>
+    </div>
+  );
+}
+
+export default function ExecuteView({
+  app,
+  user,
+  ga_tracking_id,
+  inputDataset,
+}) {
+  const { data: session } = useSession();
+
+  // TODO: check idiomatic way to do this
+  const [spec, _setSpec] = useState(JSON.parse(app.savedSpecification || `[]`));
+  const [config, __setConfig] = useState(extractConfig(spec));
+  const [inputDatasetKeys, _setInputDatasetKeys] = useState(
+    checkDatasetData(inputDataset.dataset.data, false)
+  );
+
+  const [inputData, setInputData] = useState({});
+  const isInputDataValid = () =>
+    inputDatasetKeys.every((k) => (inputData[k] || "").length > 0);
+
+  const [isRunning, setIsRunning] = useState(false);
+  const [isRunComplete, setIsRunComplete] = useState(false);
+
+  const [executionLogs, setExecutionLogs] = useState({
+    blockOrder: [],
+    lastEventByBlockName: {},
+    outputByBlockName: {},
+  });
+
+  const [outputExpandedByBlockName, setOutputExpandedByBlockName] = useState(
```

Comment:
I have the state here instead of downstream because I need to auto-expand the last output block after the run is over. It doesn't feel super right 🤔

Reply:
I need to dig a bit more to answer